### ISSUE TITLE
Unboxed empty type

### ISSUE THREAD

**Issue by treeowl** [ghc-proposals/ghc-proposals#534]
Today, we can write

```haskell
type Void :: TYPE (BoxedRep l)
data Void
```

However, there does not seem to be *any* way to produce types with other representations that GHC's pattern checker and optimizer will *recognize* as empty. For example, I can write something like

```haskell
type VoidUnit# :: TYPE (TupleRep '[])
newtype VoidUnit# = VoidUnit# VoidUnit#

absurdUnit# :: VoidUnit# -> a
absurdUnit# (VoidUnit# v) = absurdUnit# v
```

But I *can't* write

```haskell
absurdUnit# = \case
```

without annoying the pattern checker, and similarly the pattern checker won't take the type's emptiness into account in strict nested settings. The optimizer also doesn't know what's what, so it will produce an infinite loop. I can't even use a `COMPLETE` pragma to fix the pattern checking problem, because `COMPLETE` pragmas can't be empty! (Why am I not allowed to write `{-# COMPLETE :: VoidUnit# #-}`? I don't know.)

----

My first thought was that maybe there should be a magical built in `type PolyVoid# :: TYPE rep`. Now I'm thinking there's a cleaner solution. What if `UnliftedData` were expanded to allow *syntactically empty* data declarations for *any* representation type? Then we could just write

```haskell
type VoidUnit# :: TYPE (TupleRep '[])
data VoidUnit#

absurdUnit# :: VoidUnit# -> a
absurdUnit# = \case
```

What about representation polymorphic voids, and representation polymorphic absurdities? I think these are possible in principle, but they're just a bit tricky. Consider this:

```haskell
type VoidTuple# :: forall ks. TYPE (TupleRep ks)
data VoidTuple#

data Grumble ks = Grumble Groan (VoidTuple# ks)

absurdTuple# :: forall ks a. VoidTuple# @ks -> a
absurdTuple# = \case
```

`Grumble` and `absurdTuple#` are representation polymorphic in arguments, which is (currently) utterly forbidden. But ... there's actually no inherent difficulty compiling them, because they *don't need code*.

On the other hand, trying to be totally representation polymorphic runs into the problem of `LiftedRep`. So I think we'd have to make sure that any representation polymorphic empty type cannot inhabit `Type`. Or something like that.

**Comment by Jashweii** [comment]
What if GHC.Types exposed a builtin of kind ``TYPE (SumRep '[])``?

**Comment by treeowl** [comment]
> What if GHC.Types exposed a builtin of kind `TYPE (SumRep '[])`?

I think I've actually asked for that in the past, but it's insufficient. There should be recognized empty types in *every* representation. This idea was inspired by something that looked vaguely like this:

```haskell
data Foo a b c = L !(T a) b | R c
type family T (a :: Bool) :: Type where
  T True = ()
  T False = Void
```

So `Foo True b c` is like `Either b c`, and `Foo False b c` is like `Solo c`. The problem is that the `L` constructor has a field with a (useless) pointer to something of type `T a`. I'm suggesting we should be able to write something like

```haskell
data Foo a b c = L !(T a) b | R c
type family T (a :: Bool) :: TYPE (TupleRep '[]) where
  T True = (##)
  T False = VoidUnit#
```

Now that extra field will have zero width, so it's free.

**Comment by clyring** [comment]
For what it's worth, the useless pointer in your last example can also be avoided by unpacking the zero-width equality-evidence carried by a GADT constructor:

```
data Foo a b c = L {-# UNPACK #-} !(a :~: True) b | R c
```

**Comment by Ericson2314** [comment]
Just sanity-checking myself here, but isn't `TYPE (TupleRep '[])` unboxed unit not unboxed void?

**Comment by treeowl** [comment]
> Just sanity-checking myself here, but isn't `TYPE (TupleRep '[])` the kind of an unboxed unit not an unboxed void?

It is the kind of `(##)`, yes. But there are uninhabited types of every typerep.

**Comment by sgraf812** [comment]
Yes, you can write empty types at every representation through your newtype trick. It is also possible in `Type` by means of a data type with a strict field as `Abyss` in [Note \[Fuel for the inhabitation test\]](https://gitlab.haskell.org/ghc/ghc/-/blob/c189b831c74a550ddb3b94cf9b9f8922856b6990/compiler/GHC/HsToCore/Pmc/Solver.hs#L1700). As the Note explains, the problem is that the current implementation of the inhabitation test does does not detect `Abyss`/`VoidUnit#` as empty, because it has no means of inductive reasoning. It's kind of like strictness analysis, really, where we can also detect that a function always diverges. https://arxiv.org/pdf/1702.02281.pdf describes an algorithm for OCaml, IIRC. 

I had a few tries at such a test in the past, concluding that it's quite hopeless to marry it with the other tricky stuff that is happening in the checker; for example, the fuel-based approach from the Note says that we have to "give up" when we run out of fuel and say "inhabited" (which is kind of the "Top" element of the lattice). Inductive reasoning means we'd have to do some kind of fixed-point iteration starting from "Bottom" ("not inhabited") and I'm entirely unsure 

- What state flows into the fixed-point. That's what we'd have to compare on each iteration
- What happens when we run out of fuel
- Should we have a separate fuel parameter for the maximum number of fixed-point iterations?
- etc.

So I think it makes sense to want a `type Empty  :: forall (r::RuntimeRep). TYPE r` (maybe even just an empty type family recognised by the compiler) where we don't have any way to construct such a thing and the checker may presume it empty. Except for `BoxedRep Lifted`, urgh.
